defi/single-token-amm-dex/src/lib.rs

```diff
@@ -0,0 +1,914 @@
+
+            let xrd_output: Bucket = self.collected_xrd.take(self.xrd_fee);
+            
+            self.xrd_claimed += xrd_amount-self.collected_xrd.amount();
+            info!(" Protocol fee claimed {} XRD ", self.xrd_claimed);
+            
+            xrd_output
+        }
+
+            // Stock candy function callable by an end user wishing to supply unpresent liquidity to 
+            // protocol.
+        pub fn stock_candy(&mut self, candy: Bucket, new_price: Decimal) -> (Bucket,Bucket) {
+            let candy_addr = candy.resource_address();
+            let candy_amnt = candy.amount();
+            let candy_res_def = candy.resource_def();
+
+            let name = candy_res_def.metadata()["name"].clone();
+            let symbol = candy_res_def.metadata()["symbol"].clone();
+
+            assert!( candy_addr != self.collected_xrd.resource_address()," Cannot stock XRD as candy ");
+            assert!(new_price > dec!(0), "new price must be a positive value");
+            
+            match self.candy_map.get(&candy_addr) {
+                Some((_a,_b,_c)) => { 
+                    info!(" Candy already in Vault. Please use restock_candy function ");
+                    std::process::abort()
+                }
+                _ => info!(" Added {} {} candy, {} symbol @{}XRD price ", candy_amnt, name, symbol ,new_price)
+            }
+            
+            let v = self.candy_vaults.entry(candy_addr).or_insert(Vault::new(candy_addr));
+            
+            v.put(candy);
+            
+            let none: Decimal = dec!(0);
+            
+            self.candy_map.insert(candy_addr,(name.clone(),symbol.clone(),new_price));
+            self.meta_map.insert(candy_addr,(none,none,candy_addr));
+            
+            CandyDex::add_meta_candy(self, name.clone(), symbol.clone(), candy_addr);
+            
+            let meta_candy: Bucket = CandyDex::meta_mint(self, candy_amnt, candy_addr);           
+            
+            let meta_amount = meta_candy.amount();
+            let meta_badge: Bucket = 
+                CandyDex::add_meta_badge(self, symbol, candy_addr, none, meta_amount);
+            
+            self.badge_map.insert(meta_badge.resource_address().clone(),(none,none,candy_addr));
+            
+            let output_badge: Bucket = CandyDex::meta_amounts(self, meta_badge, meta_amount, none);
+            
+            (meta_candy,output_badge)
+        }
+
+            // Restock candy function callable by an end user wishing to supply present liquidity to 
+            // protocol.
+        pub fn restock_candy(&mut self, candy: Bucket) -> (Bucket,Bucket) {
+            let candy_addr = candy.resource_address();            
+            assert!( candy_addr != self.collected_xrd.resource_address()," Cannot stock XRD as candy ");
+            
+            let amnt = candy.amount();
+            
+            match self.candy_map.get(&candy_addr) {
```

Comment:
Same thing here

Reply:
Undone with objection:
Due to the fact I have to retrieve values to populate " info! ", I have to perform something like:

	if self.candy_map.contains_key(&candy_addr) {
                let (a,b,c) = self.candy_map.get(&candy_addr).unwrap();
                info!(" Adding {} {} candy, {} symbol, @{} $XRD price ", amnt, a.to_string(), b.to_string(), c);
            } else {
                info!(" Found no candy in Vault. Please use stock_candy function ");
                std::process::abort()
            }

correct me if I'm wrong it seems to me more demanding in terms of computational resources than the original one.